rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,857 @@
+configured to use an HSM or KMS for CAs.
+
+### `hsm-v1` suite
+
+Identical to `balanced-v1` suite except:
+
+* instances of Ed25519 *for CA key types only* are replaced with ECDSA with NIST P-256
+
+Subjects may still use Ed25519 keys, the HSM is only used for CA keys so the
+algorithms are limited to what the HSM can support.
+
+We claim to support and test with YubiHSM2, AWS CloudHSM, and GCP KMS.
+All of these support ECDSA with NIST P-256, none support Ed25519.
+
+This suite will *not* be compatible with clusters running in FIPS mode, because
+subjects may use Ed25519.
+In case both FIPS mode and an HSM or KMS are desired, the `fips-v1` suite should
+be used instead, it is also compatible with all HSMs and KMS we support.
+
+### `legacy` suite
+
+2048-bit RSA keys are used by all existing CAs and subjects.
+
+If new CAs or subjects are added, they should use the same choice as
+`fips-v1` where possible, because the `legacy` suite may be used in clusters
+running in FIPS mode or with an HSM.
+
+### CA details
+
+Each Teleport CA holds 1 or more of the following:
+
+* SSH public and private key
+* TLS certificate and private key
+* JWT public and private key
+
+Each CA key may be a software key stored in the Teleport backend, an HSM key
+held in an HSM connected to an Auth server via a PKCS#11 interface, or a KMS key
+held in GCP KMS.
+In the future we will likely support more KMS services.
+
+Teleport currently has these CAs:
+
+#### User CA
+
+keys: ssh, tls
+
+uses: user ssh cert signing, user tls cert signing, ssh hosts trust this CA
+
+* current/`legacy` SSH key type: 2048-bit RSA
+* proposed `balanced-v1` key type: Ed25519
+* proposed `fips-v1` key type: ECDSA with NIST P-256
+* proposed `hsm-v1` key type: ECDSA with NIST P-256
+* reasoning:
+  * Ed25519 is currently considered by multiple sources to be the best
+    algorithm for SSH
+  * Ed25519 support was added to OpenSSH 6.5 in January 2014, *before* SHA-2
+    hash support was added for RSA keys in OpenSSH 7.2
+  * ECDSA with P-256 has Go BoringCrypto support, Ed25519 doesn't
+
+* current/`legacy` TLS key type: 2048-bit RSA
+* proposed `balanced-v1` key type: ECDSA with NIST P-256
+* proposed `fips-v1` key type: ECDSA with NIST P-256
+* proposed `hsm-v1` key type: ECDSA with NIST P-256
+* reasoning:
+  * external tools may need to load user X.509 certificates signed by this CA,
+    e.g. for application access, and Ed25519 support is generally spotty
+
+#### Host CA
+
+keys: ssh, tls
+
+uses: host ssh cert signing, host tls cert signing, ssh clients trust this CA
+
+* current/`legacy` SSH key type: 2048-bit RSA
+* proposed `balanced-v1` key type: Ed25519
+* proposed `fips-v1` key type: ECDSA with NIST P-256
+* proposed `hsm-v1` key type: ECDSA with NIST P-256
+* reasoning:
+  * Ed25519 is currently considered by multiple sources to be the best
+    algorithm for SSH
+  * Ed25519 support was added to OpenSSH 6.5 in January 2014, *before* SHA-2
```

Comment:
This is missing one important point. OpenSSH is not the only SSH client available. I was fixing PyCharm client multiple times as they are using some Java client that supports almost nothing.
I think we should make sure that:
1. Users are well informed what are the consequences for this migration
2. Make sure that rolling back the changes is possible from any state.